<html><head>
  <meta content="text/html; charset=UTF-8" http-equiv="content-type">
   <style type="text/css">

   p { padding-top:2pt;padding-bottom:2pt;orphans:2;widows:2;text-align:left;
       color:#222;font-weight:400;text-decoration:none;vertical-align:baseline;
       font-size:12pt;line-height:135%;font-family:"Arial";font-style:normal }
   ol{margin:4px;padding:4px}
   li{margin:2px;padding:2px}
   table td,table th{padding:0}
   .c1{color:#a10036;font-weight:700;text-decoration:none;vertical-align:baseline;font-size:16pt;font-family:"Arial";font-style:normal}
   .c6{padding-top:18pt;padding-bottom:6pt;line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}
   .c4{color:#222;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:12pt;line-height:140%;font-family:"Arial";font-style:normal}
   .c0{padding-top:0pt;padding-bottom:2pt;line-height:140%;orphans:2;widows:2;text-align:left}
   .cbody {font-size:12pt;background-color:#ffffff;max-width:824px;padding:36pt 36pt 36pt 72pt}
   .bred{font-weight:bold;color:#b32141;}
   .red{font-weight:normal;color:#b32141;}
   .c5{font-size:12pt;}


   .title{padding-top:0pt;color:#666;font-size:20pt;font-family:"Arial";page-break-after:avoid;orphans:2;widows:2;text-align:left}
   .subtitle{padding-top:0pt;color:#666666;font-size:15pt;padding-bottom:16pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}
   li{color:#000000;font-family:"Arial"}
   h1{padding-top:20pt;color:#222;font-size:20pt;padding-bottom:2pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}
   h2{padding-top:12pt;color:#222;font-size:16pt;padding-bottom:1pt;font-family:"Arial";page-break-after:avoid;orphans:2;widows:2;text-align:left}
   h3{padding-top:6pt;color:#434343;font-size:14pt;padding-bottom:1pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}
   h4{padding-top:4pt;color:#423c3c;font-size:12pt;padding-bottom:1pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}
   h5{padding-top:12pt;color:#666666;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}
   h6{padding-top:12pt;color:#666666;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;font-style:italic;orphans:2;widows:2;text-align:left}
   </style>
   <title>LIBData Portal Guide</title>
  </head>
  <body class="cbody">
     <!-- h2 class="title"><span class="bred">L</span>ieber <span class="bred">I</span>nstitute
         <span class="bred">D</span>ata <span class="bred">Portal</span></h2 -->
    <h2 class="title" style="margin-bottom:0;padding-bottom:0;"><span class="bred">LIBD Data Portal</span> guide</h2>
    <h3>Logging in</h3>
    <p>A login dialog is presented at the start of the application, allowing user authentication using the LIBD email (username only, no need for
      <i>@libd.org</i>) and password.<br/><br/>
     Authentication is needed in order to export/download data from the application. The login dialog can be dismissed without loggin in,
     in which case data selection and exploration are still possible, but actual data exports (including genotypes) will not be available
     unless the user is authenticated. <br/><br/>
     Logging in can take place at any point during/after the data selection (using the <span class="red">Login</span>
     link at the top-right side of the header), it does not affect the current selection, and it should immediately enable the data export/request
     buttons in the application.
     <br/><br/>
      <b>Note:</b>&nbsp;in some browsers, when the user is filling the login form, there might be a warning displayed about
      login security, stating that the connection "is not secure". This is not accurate because even though the main web
      application is running on an unecrypted connection, the login form itself is using a different  connection which is
      secure and encrypted (https). Thus the password entered in that dialog is relayed to an authentication server using
      only encrypted connections and it is never passed to the main web application.
     </p>

    <h3 id="brsel-matrix">General navigation</h3>
    <p>   This is a multi-page web application, please do not use the <i>Back</i> button (&#x2190;) of the browser. Instead, in order to
      navigate between pages, the active page or section of the web app should be selected using the top-left navigation button:
    </p>
    <p>
        <span style="overflow: hidden; display: inline-block; margin: 6px 2px; border: 0.00px solid #000000;
          transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px); width: 356.00px; height: 115.00px;">
          <img alt="" src="images/image1.png" style="width: 356.00px; height: 115.00px; margin-left: 0.00px; margin-top: 0.00px; transform: rotate(0.00rad) translateZ(0px); -webkit-transform: rotate(0.00rad) translateZ(0px);" title="">
        </span>
    </p>
    <p>
          This is a button which when clicked displays a drop-down menu with the available application pages.
          The pages not available yet are grayed out.
    </p>

    <p>
          The entry page is the <span class="bred">Brain Set Builder</span>,
            however if the user
            is just interested in available datasets for specific data types (e.g. <span class="bred">Bulk RNAseq </span> data),
            the user can navigate directly to the data type of interest by using this top-left navigation button.
    </p>

         <h2><span class="c6 bred">Brain Set Builder</span> page</h2>
          <p>The <span class="bred">Brain Set Builder</span>
            page allows the user to get an overview of the available data types for the subjects loaded in the database,
            and make an informed selection of a set of subjects of interest based on diagnosis and demographics as well as availability
            of experiment data for the subjects.
          </p>
          <p>
            The <span class="bred">Brain Set Builder</span> page has two sections (tabs): <span class="red">Select</span> and
            <span class="red">Browse</span>, selectable to the right of the main page navigation button:<br/><br/>
            <span style="overflow: hidden; display: inline-block; margin: 6px 2px;" >
              <img alt="" src="images/brain_set_builder-tabs.png" style="margin-left: 0.00px; margin-top: 0.00px;" title="">
            </span>
          </p>

          <h3>The <span class="bred">Select</span> tab</h3>
          <p>
          The main tab of the <span class="red">Brain Set Builder</span> page allows inspection and selection of subjects by demographics/phenotype data,
          brain region assay data availability, or a user provided list of brain identifiers.
          </p>
          <h4>Subject selection by phenotype</h4>
          <p>In order to create a set of subjects a selection must be made by clicking to highlight one or more categories in the available panels
            and then clicking the <span class="bred" style="color:#c34">Apply</span> button that appears at the top of a category panel
              whenever the highlighted criteria have changed for that panel.
              <div style="overflow: hidden; width:100%; display: block; margin: 6px auto;" >
                 <img alt="" src="images/apply-before.png" style="display:block;margin: 0 auto;" title="Selection not applied yet">
              </div>
          </p>
          <p> Only after clicking the <span style="color:#c34"><b>Apply</b></span> button the selection will actually
            become effective and the counts in the other category panels will be updated to reflect the applied selection.
            The current selection is applied (effective) if the <span style="color:#c34"><b>Apply</b></span> button is
            no longer visible and the selected categories are displayed in a higher contrast:

            <div style="overflow: hidden; width:100%; display: block; margin: 6px auto;" >
              <img alt="" src="images/apply-after.png" style="display:block;margin: 0 auto;" title="Applied selection">
            </div>

          <p>Note that applying a selection in a panel does not change its own displayed counts - such that the user can still see the potential selection
            counts for the other categories in that panel.</p>
          </p>

            <h4>Region & Data type selection</h4>

          <table><tr>
          <td style="vertical-align:top;">
          <span style="overflow: hidden; display: inline-block; margin: 6px 2px;" >
            <img alt="" src="images/matrix_selection.png" style="margin-left: 0.00px; margin-top: 0.00px;" title="">
          </span>
          </td><td style="vertical-align:top;padding-left:6px;">
          <p>
            The table in the center of the page shows how many subjects were sequenced by various data types across brain regions, as loaded in the database.
            <br /><br />A subject can have samples sequenced from multiple brain regions, or different experimental assays,
            so they are counted independently per brain region and assay data type in this table.
            <br /><br /> Selecting the first region cell locks the selection to that data type (column). The selection of region cells in this matrix
            will restrict the subjects to only those that have sequencing data for the selected brain regions, for the selected data type (column).
          </p>
          </td>
         </tr></table>

         <h4>Selection by list of brain identifiers</h4>
         <p>Another way to select subjects and inspect their information is to provide a list of subjects using their brain identifiers.
         The [<b>Load Br# list</b>] button opens a dialog which allows uploading or directly entering a list of "brain numbers"
         in the usual LIBD brain identifier format (e.g. <b><code>Br5362</code></b>). A text file that has such <i>Br####</i> identifiers can be uploaded by
         the user (including full CSVs with many other data columns, like RNAseq samples metadata exported from the
         <span class="bred">Bulk RNAseq</span> page), or text that can be directly entered/pasted in the text input area of that dialog.<br/><br/>
         Provided Br#s will be verified against the internal database and only the matching ones will be considered.
         <br/><br/>

         Please note that upon providing a custom Br# list this way, the current selection filters <b>will be cleared</b>
         and all the category panels will be updated to reflect the counts as constrained to the list of Br#s provided.<br/>
         However the selection can then be further refined by applying further selections in any of the panels.
         </p>

        <h4 id="brsel-browse">The Brain Set <span style="color:#c34">Browse</span> tab</h4>
        <p>
        The <span style="color:#c34">Browse</span> tab of the <b>Brain Set Builder</b> allows the inspection of the subject data for all the selected subjects,
        and exporting the resulting table as a CSV/TSV file.
        It also allows further customization of the subject data table based on the available sequencing information for each brain.
        <br /><br />
        The default table shows the demographics data along with the total number of sequencing samples for each experiment data type, as loaded in the database.
        <br /><br />
        The checkbox labeled <b>Show sample counts by brain region</b> above the table allows switching to a detailed view of sample counts by brain region, so the table will expand accordingly
        to show all the sequenced brain regions and the sample count per experiment data type within each region.
        <br /><br />
        The subject set table can be customized to include/exclude the columns specific to experiment data types (the first set of checkboxes, which are all checked by default),
        but also to further restrict the subject set to only those brains that have specific experiment data type available (the second set of checkboxes on the right).
        <br />
        <br />
        Note that in this  <span style="color:#c34">Browse</span> tab, the subject selection can be further refined (filtered)
        by using the list of checkboxes on the right:
        <div style="overflow: hidden; width:100%; display: block; margin: 6px auto;" >
          <img alt="" src="images/browse-brain-filter.png" style="display:block;margin: 0 auto;" title="Applied selection">
       </div><br/>
       <p>
       If any entry is checked in this list, a subject shall only be included in the selection when it has experiment data
       available for all data types that are checked in that list.
          <br/><br/>
          Finally, use the <b>Export Set</b> button to save a CSV/TSV with the current subject selection and their metadata, or use the
          <b>Request Genotypes</b> button to request a VCF.gz file with the imputed genotypes for the selected subjects.<br/><br/>

          When choosing the "Request Genotypes" button and confirming the request, a custom VCF.gz file with the imputed genotype data of the
          selected subjects will begin assembling in the background on the database server. This can be a lengthy operation (15-20 minutes or more),
          depending on the number of subjects and server load. The user can continue to use the web application normally while the genotypes file
          is being built.
          <br/>The resulting genotypes file can be up to several gigabytes in size. When the file is ready, a direct download
          link to the custom genotypes file will be e-mailed to the user. Please note that the genotypes download link expires in 3 days after receiving
          the notification email.
         </p>
       </p>

        <h2 id="rna-sel"><span class="c6 bred">Bulk RNAseq</span> page</h4>
        <p>
          On this page the user can make selections of RNAseq expression values by subsetting the available bulk RNAseq datasets
          pre-processed and stored in the database.<br/><br/>

          The RNAseq datasets shown on this page were processed using <a href="https://www.gencodegenes.org/human/release_25.html">Gencode Release 25</a>
           as the reference annotation, using the <a href="https://doi.org/10.1186/s12859-021-04142-3">SPEAQeasy pipeline</a> <br />
          <br/>
          Expression values (read counts) were obtained with these methods:
          <ul>
           <li><b>gene</b> and <b>exon</b> read counts were obtained with <a href="https://doi.org/10.1093/bioinformatics/btt656">featureCounts</a>.</li>
           <li><b>junction</b> read counts were obtained with <a href="https://www.biorxiv.org/content/10.1101/436634v5">regtools</a>.</li>
           <li><b>transcript</b> read counts were estimated and normalized as TPM values using <a href="https://www.nature.com/articles/nbt.3519">kallisto</a>
            or <a href="https://pubmed.ncbi.nlm.nih.gov/28263959/">Salmon</a>.</li>
          </ul>

          <br/>
          <p>
            The <span class="bred">Bulk RNAseq</span> page has two sections (tabs):
            <span class="red">Select</span> and <span class="red">Explore</span>, selectable to the right of the main page navigation button:<br/>
            <span style="overflow: hidden; display: inline-block; margin: 6px 2px;" >
              <img alt="" src="images/bulk_rnaseq-tabs.png" style="margin-left: 0.00px; margin-top: 0.00px;" title="">
            </span>
          </p>

          <h4> Bulk RNAseq <span style="color:#c34">Select</span> tab</h4>
         <p>
          Please note that the numbers (counts) displayed in each selection panel represent the <b>number of RNAseq experiments (samples)</b>,
          not subject counts. Similar to the way selection works in the <span class="bred">Brain Set Builder</span> UI, applying a selection in any panel will
          update the counts in any other panel (but not its own).<br /><br/>

          Similarly the <b>Load Br# list</b> button can be used to load a custom set of <i>Br####</i> identifiers -- for example a CSV saved with
          the <b>Export set</b> button in <span class="bred">Brain Set Builder</span>). Just like on that page, loading a custom Br# list
          will clear the existing filters in order to reflect the distribution of the provided Br#s in all the category panels,
          but the selection can then be further refined by applying selections in any of the panels.
          <br/><br/>

          At least a dataset and a primary diagnosis must be selected before the user can gain access to the samples for the selection (i.e.
          to show the "Export" and the "Explore" buttons)
          <br/><br/>
          Selecting multiple datasets at once is discouraged (due to batch effect considerations, different sequencing protocols etc.).
          For that reason the <b>Datasets</b> panel is the only list where single selection is the default, i.e. simply selecting a list item
          (dataset) in that panel cancels the previous selection (as usual, the new selection still has to be "applied"). <br/><br/>
          However multiple datasets can still be selected by using <i>Ctrl</i>+click on multiple datasets in that panel, thus allowing
          RNAseq sample selection <i>across</i> datasets.
        </p>
        <h4>Publication info</h4>
        <p>
        Some of the bulk RNAseq datasets that have been already published have reference information and other dataset-specific related links (e.g. a link to the
        the LIBD EQTL browser for that dataset). A panel with this reference information is displayed above
        the "Datasets" panel when such a dataset is selected in that list.
        <div style="overflow: hidden; width:100%; display: block; margin: 6px auto;" >
          <img alt="" src="images/reference_publication.png" style="display:block;margin: 0 auto;" title="Applied selection">
        </div><br/>
        </p>

        <h4 id="rna-explore">Bulk RNAseq <span style="color:#c34">Explore</span> tab</h4>
        <p>
        On the <span style="color:#c34;">Explore</span> tab (also activated when the <i>"Explore"</i> button is clicked for a valid selection),
        a few basic visualisations (plots) can be displayed for the current selection (which is again summarized at the top).
        <br/><br/>
        Use the side bar menus on the left to select the plot type and enter the gene names for which the expression data shall be plotted.
        </p>


  </body>
  </html>